crates/bevy_remote/src/lib.rs

```diff
@@ -0,0 +1,856 @@
+//! Arbitrary system parameters can be used in conjunction with the optional `Value` input. The
+//! handler system will always run with exclusive `World` access.
+//!
+//! [the `serde` documentation]: https://serde.rs/
+
+#![cfg(not(target_family = "wasm"))]
+
+use std::{
+    net::{IpAddr, Ipv4Addr},
+    sync::RwLock,
+};
+
+use anyhow::Result as AnyhowResult;
+use bevy_app::prelude::*;
+use bevy_derive::{Deref, DerefMut};
+use bevy_ecs::{
+    entity::Entity,
+    system::{Commands, IntoSystem, Res, Resource, System, SystemId},
+    world::World,
+};
+use bevy_reflect::Reflect;
+use bevy_tasks::IoTaskPool;
+use bevy_utils::{prelude::default, HashMap};
+use http_body_util::{BodyExt as _, Full};
+use hyper::{
+    body::{Bytes, Incoming},
+    server::conn::http1,
+    service, Request, Response,
+};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use smol::{
+    channel::{self, Receiver, Sender},
+    Async,
+};
+use smol_hyper::rt::{FuturesIo, SmolTimer};
+use std::net::{TcpListener, TcpStream};
+
+pub mod builtin_methods;
+
+/// The default port that Bevy will listen on.
+///
+/// This value was chosen randomly.
+pub const DEFAULT_PORT: u16 = 15702;
+
+/// The default host address that Bevy will use for its server.
+pub const DEFAULT_ADDR: IpAddr = IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1));
+
+const CHANNEL_SIZE: usize = 16;
+
+/// Add this plugin to your [`App`] to allow remote connections to inspect and modify entities.
+///
+/// The defaults are:
+/// - [`DEFAULT_ADDR`] : 127.0.0.1.
+/// - [`DEFAULT_PORT`] : 15702.
+pub struct RemotePlugin {
+    /// The address that Bevy will use.
+    address: IpAddr,
+
+    /// The port that Bevy will listen on.
+    port: u16,
+
+    /// The verbs that the server will recognize and respond to.
+    methods: RwLock<Vec<(String, Box<dyn System<In = Option<Value>, Out = BrpResult>>)>>,
+}
+
+/// A resource containing the IP address that Bevy will host on.
+#[derive(Resource)]
+pub struct HostAddress(pub IpAddr);
+
+/// A resource containing the port number that Bevy will listen on.
+#[derive(Resource, Reflect)]
+pub struct HostPort(pub u16);
+
+/// The type of a function that implements a remote method (`bevy/get`, `bevy/query`, etc.)
+///
+/// The first parameter is the JSON value of the `params`. Typically, an
+/// implementation will deserialize these as the first thing they do.
+///
+/// The returned JSON value will be returned as the response. Bevy will
+/// automatically populate the `id` field before sending.
+pub type RemoteMethod = SystemId<Option<Value>, BrpResult>;
+
+/// Holds all implementations of methods known to the server.
+///
+/// Custom methods can be added to this list using [`RemoteMethods::insert`].
+#[derive(Resource, Default)]
+pub struct RemoteMethods(HashMap<String, RemoteMethod>);
+
+/// A single request from a Bevy Remote Protocol client to the server,
+/// serialized in JSON.
+///
+/// The JSON payload is expected to look like this:
+///
+/// ```json
+/// {
```

Comment:
Ah. It does do that; this was just an un-updated doc comment.

Reply:
Now that I look at the code again, it looks like we also didn't update this to actually check the incoming `jsonrpc` version number for requests; I had just always been testing with `"jsonrpc": "2.0"` which was swiftly ignored. I guess we didn't catch this earlier because of how `serde` deserializes json maps into structs.

Reply:
In addition to updating the docs I belive the `BrpRequest` struct itself should have this field

```rust
    /// This field is mandatory and must be set to `"2.0"`.
    pub jsonrpc: &'static str,
```

copied from `BrpResponse`

Reply:
Looking at the compatibility section of the spec:

> JSON-RPC 2.0 Request objects and Response objects may not work with existing JSON-RPC 1.0 clients or servers. However, it is easy to distinguish between the two versions as 2.0 always has a member named "jsonrpc" with a String value of "2.0" whereas 1.0 does not. Most 2.0 implementations should consider trying to handle 1.0 objects

I suppose not caring if clients put the `2.0` thing in requests is a way of handling v1 clients, if that's the case then maybe leaving it out of the struct isn't a bad idea.